File: day05/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = true;
// const DEBUG = false;

const data = await Deno.readTextFile('input.txt');
if (DEBUG) console.debug(data);
const sections = data.split('\n\n');
const rules = sections[0]
  .split('\n')
  .filter((line) => line.trim() != '')
  .map((line) => line.split('|').map((item) => parseInt(item)));
if (DEBUG) console.debug({ rules });
const updates = sections[1]
  .split('\n')
  .filter((line) => line.trim() != '')
  .map((line) => line.split(',').map((item) => parseInt(item)));
if (DEBUG) console.debug({ updates });

let total1 = 0;
const unsortedUpdates: number[][] = [];
updates.forEach((update) => {
  const correctOrder = rules.every(
    (rule) => !(update.includes(rule[0]) && update.includes(rule[1])) || update.indexOf(rule[0]) < update.indexOf(rule[1])
  );
  if (DEBUG && correctOrder) console.debug({ update, correctOrder });
  if (correctOrder) {
    total1 += update[Math.round((update.length - 1) / 2)];
  } else {
    unsortedUpdates.push(update);
  }
});
console.log('part 1', total1);

let total2 = 0;
unsortedUpdates.forEach((update) => {
  const updateRules = rules.filter((rule) => update.includes(rule[0]) && update.includes(rule[1]));
  if (DEBUG) console.debug({ update, updateRules });
  let moved = true;
  while (moved) {
    moved = false;
    updateRules.forEach((rule) => {
      const leftIx = update.indexOf(rule[0]);
      const rightIx = update.indexOf(rule[1]);
      if (leftIx > rightIx) {
        if (DEBUG) console.debug('move', { rule, leftIx, rightIx });
        update.splice(leftIx, 1);
        update.splice(rightIx > 0 ? rightIx - 1 : rightIx, 0, rule[0]);
        moved = true;
      }
    });
  }
  total2 += update[Math.round((update.length - 1) / 2)];
});
console.log('part 2', total2);

File: day05/solution.py
#!/usr/bin/env python3

with open('input.txt', 'r') as f:
  data = f.read()

sections = data.split('\n\n')
rules = [list(map(int, x.split('|'))) for x in sections[0].splitlines()]
updates = [list(map(int, x.split(','))) for x in sections[1].splitlines()]

# print(f'{rules=}')
# print(f'{updates=}')

sum_middle = 0
unsorted_updates = []
for update in updates:
  correct_order = True
  for rule in rules:
    if rule[0] in update and rule[1] in update:
      if update.index(rule[0]) > update.index(rule[1]):
        correct_order = False
        break
  if correct_order:
    sum_middle += update[int((len(update) - 1) / 2)]
  else:
    unsorted_updates.append(update)
print(f'part 1 {sum_middle=}')

sum_middle = 0
for update in unsorted_updates:
  orig = [*update]
  moved = True
  while moved:
    moved = False
    for rule in rules:
      if rule[0] in update and rule[1] in update:
        if update.index(rule[0]) > update.index(rule[1]):
          temp = [x for x in update if x != rule[0]]
          temp.insert(temp.index(rule[1]), rule[0])
          print(f'move {rule[0]} before {rule[1]}')
          update = temp
          moved = True
  print(f'{orig=} {update=}')
  sum_middle += update[int((len(update) - 1) / 2)]
print(f'part 2 {sum_middle=}')
